### PR TITLE
Refactor WorkingBeatmap

### PR DESCRIPTION
**Pull request by ddevault**
Gets ArchiveReader out of WorkingBeatmap and delegates extracting stuff
from it to subclasses. Should enable us to make an OnlineWorkingBeatmap
or so.

### CODE REVIEW

File: osu.Desktop.VisualTests/Beatmaps/TestWorkingBeatmap.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        public override Texture Background => null;
+        public override Track Track => null;
+        
+        public override void Dispose()
+        {
+            // This space intentionally left blank
+        }
+        
+        public override void TransferTo(WorkingBeatmap other)
+        {
```
Comment (SirCmpwn) [inline osu.Desktop.VisualTests/Beatmaps/TestWorkingBeatmap.cs]: Could possibly avoid making subclasses implement this by giving Track a protected setter

File: osu.Game/Beatmaps/WorkingBeatmap.cs [hunk 2]
```diff
@@ -27,131 +24,26 @@ public abstract class WorkingBeatmap : IDisposable
         /// </summary>
         public PlayMode? PreferredPlayMode;
 
-        public PlayMode PlayMode => beatmap?.BeatmapInfo?.Mode > PlayMode.Osu ? beatmap.BeatmapInfo.Mode : PreferredPlayMode ?? PlayMode.Osu;
+        public PlayMode PlayMode => Beatmap?.BeatmapInfo?.Mode > PlayMode.Osu ? Beatmap.BeatmapInfo.Mode : PreferredPlayMode ?? PlayMode.Osu;
```
Comment (SirCmpwn) [inline osu.Game/Beatmaps/WorkingBeatmap.cs]: This subtly changes the behavior of WorkingBeatmap.PlayMode but I'm pretty sure it isn't a meaningful difference

File: osu.Game/Database/DatabaseWorkingBeatmap.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Formats;
+using osu.Game.Beatmaps.IO;
+
+namespace osu.Game.Database
+{
+    public partial class BeatmapDatabase
+    {
+        private class DatabaseWorkingBeatmap : WorkingBeatmap
+        {
```
Comment (SirCmpwn) [inline osu.Game/Database/DatabaseWorkingBeatmap.cs]: Discussion point, made this a partial class just to move the nested private class to its own file

**Comment by peppy** [comment]
not sure the caching logic should be moved out of the base `WorkingBeatmap` class (for things like `Track`/`Background`). The whole idea of a `WorkingBeatmap` was to ensure we don't fetch the same resource more than once.

Would much rather to see `protected abstract GetTrack()` functions but leave the caching in the base class.

**Comment by SirCmpwn** [comment]
Good idea. Do you think the locking should be moved back to the parent class then as well?

**Comment by peppy** [comment]
yep, take as much implementation work away from derived classes as possible.

File: osu.Game/Beatmaps/WorkingBeatmap.cs (hunk 2 above)
Comment (peppy) [inline osu.Game/Beatmaps/WorkingBeatmap.cs]: this is actually more correct

File: osu.Game/Database/DatabaseWorkingBeatmap.cs (hunk 3 above)
Comment (peppy) [inline osu.Game/Database/DatabaseWorkingBeatmap.cs]: i'm against this unless the filename follows partial conventions `BeatmapDatabase_DatabaseWorkingBeatmap.cs` ... which feels a bit too long.

Comment (SirCmpwn) [inline osu.Game/Database/DatabaseWorkingBeatmap.cs]: So would you prefer it become a public class? Or just internal? I'm not to thrilled about internal in this context

**Pull request merged**